Mathlib/NumberTheory/NumberField/CMField.Lean

```diff
@@ -0,0 +1,158 @@
+import Mathlib.NumberTheory.NumberField.Units.DirichletTheorem
+
+/-!
+# CM-extension of number fields
+
+A CM-extension `K/F` of number fields is an extension where `K` is totally complex, `F` is
+totally real and `K` is a quadratic extension of `F`. In this situation, the totally real
+subfield `F` is (isomorphic to) the maximal real subfield of `K`.
+
+## Main definitions
+
+* `NumberField.CMExtension.complexConj`: the complex conjugation of a CM-extension.
+
+* `NumberField.CMExtension.isConj_complexConj`: the complex conjugation is the conjugation
+  of any complex embedding of a `CM`-extension.
+
+* `NumberField.IsCMField`: A predicate that says that a number field is CM, that is it a totally
+  complex quadratic extension of its totally real subfield
+
+## Implementation note
+
+Most result are proved under the general hypothesis: `K/F` quadratic extension of number fields
+with `F` totally real and `K` totally complex and then, if relevant, we deduce the special case
+`F = maximalRealSubfield K`. Results of the first kind lives in the `NumberField.CMExtension`
+namespace whereas results of the second kind lives in the `NumberField.IsCMField` namespace.
+
+-/
+
+open NumberField ComplexEmbedding InfinitePlace Algebra
+
+open scoped ComplexConjugate
+
+namespace NumberField
+
+namespace CMExtension
+
+variable (F K : Type*) [Field F] [NumberField F] [IsTotallyReal F] [Field K] [NumberField K]
+  [IsTotallyComplex K] [Algebra F K] [IsQuadraticExtension F K]
+
+theorem card_infinitePlace_eq_card_infinitePlace :
+    Fintype.card (InfinitePlace F) = Fintype.card (InfinitePlace K) := by
+  rw [card_eq_nrRealPlaces_add_nrComplexPlaces, card_eq_nrRealPlaces_add_nrComplexPlaces,
+    IsTotallyComplex.nrRealPlaces_eq_zero K, IsTotallyReal.nrComplexPlaces_eq_zero F, zero_add,
+    add_zero, ← IsTotallyReal.finrank, ← Nat.mul_left_cancel_iff zero_lt_two,
+    ← IsTotallyComplex.finrank, ← Module.finrank_mul_finrank ℚ F K, mul_comm,
+    IsQuadraticExtension.finrank_eq_two F K]
+
+theorem units_rank_eq_units_rank :
+    Units.rank F = Units.rank K := by
+  rw [Units.rank, Units.rank, card_infinitePlace_eq_card_infinitePlace F K]
+
+variable {K}
+
+theorem exists_isConj (φ : K →+* ℂ) :
+    ∃ σ : K ≃ₐ[F] K, IsConj φ σ :=
+  exists_isConj_of_isRamified <|
+    isRamified_iff.mpr ⟨IsTotallyComplex.isComplex _, IsTotallyReal.isReal _⟩
+
+omit [IsTotallyReal F] in
+/--
+All the conjugations of a `CM`-extension are the same.
+-/
+theorem isConj_eq_isConj {φ ψ : K →+* ℂ} {σ τ : K ≃ₐ[F] K} (hφ : IsConj φ σ) (hψ : IsConj ψ τ) :
+    σ = τ := by
+  have : Fintype.card (K ≃ₐ[F] K) = 2 :=
+    (IsQuadraticExtension.finrank_eq_two F K) ▸ IsGalois.card_aut_eq_finrank F K
+  rw [← Nat.card_eq_fintype_card, Nat.card_eq_two_iff' 1] at this
+  exact ExistsUnique.unique this (IsTotallyComplex.isConj_ne_one hφ)
+    (IsTotallyComplex.isConj_ne_one hψ)
+
+variable (K) in
+/--
+The complex conjugation of a `CM`-extension.
+-/
+noncomputable def complexConj : K ≃ₐ[F] K :=
```

Comment:
Would it make sense to create an instance of `StarRing`, just like for the complex numbers?
Maybe in a follow-up PR?

Reply:
No, you're right. It would make more sense in the general case

Reply:
Ok, I added the instance but indeed only in the `IsCMField` case since Lean cannot figure out the maximal real subfield otherwise.